### ISSUE TITLE
Toml.FromModel fails with InvalidCastException in .net 8

### ISSUE THREAD

**Issue by TylerReid** [xoofx/Tomlyn#76]
I was updating an application to .net 8 and found a new exception happening in `Toml.FromModel` 
Reproducible example
```csharp
using Tomlyn;
Console.WriteLine(Toml.FromModel(new Dictionary<string, Test>()));
class Test{}
```
which fails with the following stack trace given `<TargetFramework>net8.0</TargetFramework>` but works in `<TargetFramework>net7.0</TargetFramework>`

```
Unhandled exception. System.InvalidCastException: Unable to cast object of type 'System.GenericEmptyEnumerator`1[System.Collections.Generic.KeyValuePair`2[System.String,Test]]' to type 'System.Collections.IDictionaryEnumerator'.
   at Tomlyn.Model.Accessors.DictionaryDynamicAccessor.SlowDictionaryAccessor.GetElements(Object dictionary)+MoveNext()
   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)
   at Tomlyn.Model.ModelToTomlTransform.VisitObject(ObjectDynamicAccessor accessor, Object currentObject, Boolean inline)
   at Tomlyn.Model.ModelToTomlTransform.Run()
   at Tomlyn.Toml.TryFromModel(Object model, TextWriter writer, DiagnosticsBag& diagnostics, TomlModelOptions options)
   at Tomlyn.Toml.TryFromModel(Object model, String& modelAsToml, DiagnosticsBag& diagnostics, TomlModelOptions options)
   at Tomlyn.Toml.FromModel(Object model, TomlModelOptions options)
   at Program.<Main>$(String[] args) in C:\Users\treid\k\tomltest\Program.cs:line 2
```

`GenericEmptyEnumerator` is a new optimized type added in .net 8. I am not 100% sure yet what the underlying cause is, but I suspect it is breaking the assumption of what an empty dictionary's enumerator will be.

I believe the change that introduced this type is here https://github.com/dotnet/runtime/pull/82899

**Comment by TylerReid** [comment]
I think one solution would be to do this at the top of `SlowDictionaryAccessor.GetElements`
```csharp
if (dictionary is IDictionary d && d.Count == 0)
{
    yield break;
}
```
or maybe change it to a pattern match before using the enumerator
```charp
        public override IEnumerable<KeyValuePair<string, object?>> GetElements(object dictionary)
        {
            var it = (IEnumerable)dictionary;
            if (it.GetEnumerator() is IDictionaryEnumerator enumerator)
            {
                while (enumerator.MoveNext())
                {

                    if (!_context.TryConvertValue(new SourceSpan(), enumerator.Key, typeof(string), out var newKey) || newKey is not string text)
                    {
                        _context.Diagnostics.Error(new SourceSpan(), $"Unable to convert key {enumerator.Key} to a string");
                        yield break;
                    }

                    yield return new KeyValuePair<string, object?>(text, enumerator.Value);
                }
            }
        }
```
but neither one of these are completely satisfying to me.

**Comment by TylerReid** [comment]
I did a bit more digging and the actual source of the error is this https://github.com/dotnet/runtime/pull/82499 because it replaced the `IEnumerable` specific `GetEnumerator` with a special case for 0 sized dictionaries. The original PR I linked was an additional optimization that leads to the type we see it fail with now. The correct solution I believe is to not cast to an `IEnumerable` and instead go through `IDictionary` so we can be sure that the enumerator is the version we need. I will open a PR for this.
